### PR TITLE
Auto register mapper classes by namespace

### PR DESCRIPTION
**Pull request by redbaty**
Title says it all.

### Usage
```csharp
using (var stream = File.OpenRead(FILEPATH))
using (var importer = new ExcelImporter(stream))
{
    importer.RegisterMapperClassesByNamespace("NAMESPACE");
    return importer.ReadSheet().ReadRows<T>().ToArray();
}
```

**(This have broken the tests)**

### CODE REVIEW

**Comment by redbaty** [comment]
The test fails with a missing "Main" entry point, since I'm not familiar XUnit help is appreciated

**Comment by hughbe** [comment]
Interesting proposition - couple of questions/comments:
- What API is netstandard2.0? Is there a way to do the same thing without having to have multiple targets?
- Is the namespace `ExcelMapper.Utilities` visible - ideally the API could be discovered without having to look use the namespace first? Maybe just move the extension into the `ExcelMapper` namespace? I see the `Utilities` sub directory as being for internal utilities rather than public APIs.

File: src/Utilities/ExcelImporterUtils.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+using System.Linq;
+using System.Reflection;
+
+namespace ExcelMapper.Utilities
+{
+    public static class ExcelImporterUtils
+    {
+#if NETSTANDARD2_0
+        public static List<ExcelClassMap> RegisterMapperClassesByNamespace(this ExcelImporter importer,
+            string namespacestr)
```
Comment (hughbe) [inline src/Utilities/ExcelImporterUtils.cs]: I don't like this param name. Could you change it to `namespace`?

Comment (hughbe) [inline src/Utilities/ExcelImporterUtils.cs]: This also needs parameter validation.

```
if (namespace== null)
{
    throw new ArgumentNullException(nameof(namespace));
}

if (namespace.Length == 0)
{
    throw new ArgumentException("The namespace cannot be empty.", nameof(namespace));
}
```

**Comment by redbaty** [comment]
@hughbe Well netstandard2.0 is the newest version of netstandard, and it's needed because it supports the Assembly type.

And I'll move it to the ExcelMapper namespace.

File: src/Utilities/ExcelImporterUtils.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace ExcelMapper.Utilities
+{
+    public static class ExcelImporterUtils
+    {
+#if NETSTANDARD2_0
+        public static List<ExcelClassMap> RegisterMapperClassesByNamespace(this ExcelImporter importer,
```
Comment (hughbe) [inline src/Utilities/ExcelImporterUtils.cs]: Please return `IEnumerable<ExcelClassMap>`

File: src/Utilities/ExcelImporterUtils.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+            string namespacestr)
+        {
+            var classes = Assembly.GetExecutingAssembly()
+                .GetTypes().AsParallel()
+                .Where(t => t.IsClass && t.Namespace == namespacestr);
+
+            var objects = classes.AsParallel().Select(Activator.CreateInstance)
+                .AsParallel().OfType<ExcelClassMap>().ToList();
```
Comment (hughbe) [inline src/Utilities/ExcelImporterUtils.cs]: You can remove the `ToList` call as this method should return `IEnumerable` instead of `List`

Comment (hughbe) [inline src/Utilities/ExcelImporterUtils.cs]: Are all these calls to `AsParallel` necessary? Surely only one is necessary? I'm not sure

File: src/Utilities/ExcelImporterUtils.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+{
+    public static class ExcelImporterUtils
+    {
+#if NETSTANDARD2_0
+        public static List<ExcelClassMap> RegisterMapperClassesByNamespace(this ExcelImporter importer,
+            string namespacestr)
+        {
+            var classes = Assembly.GetExecutingAssembly()
+                .GetTypes().AsParallel()
+                .Where(t => t.IsClass && t.Namespace == namespacestr);
```
Comment (hughbe) [inline src/Utilities/ExcelImporterUtils.cs]: Why not just do `t.IsAssignableFrom(typeof(ExcelClassMap)) && t.Namespace == namespace`

**Comment by hughbe** [comment]
The only netstandard2.0 API is `Assembly.GetExecutingAssembly`. Could you add another overload:

```
#if NETSTANDARD2_0
public static List<ExcelClassMap> RegisterClassMapsInNamespace(this ExcelImporter importer, string namespace)
{
    importer.RegisterClassMapsInNamespace(Assembly.GetExecutingAssembly(), namespace);
}
#endif

public static List<ExcelClassMap> RegisterClassMapsInNamespace(this ExcelImporter importer, Assembly assembly, string namespace)
{
    // ...
}
```

This way netstandard1.4 users can still use this API. For example by using `typeof(Program).Assembly`

File: src/Utilities/ExcelImporterUtils.cs (hunk 3 above)
Comment (redbaty) [reply src/Utilities/ExcelImporterUtils.cs]: The ToList() prevents it from enumerating multiple times

Comment (redbaty) [reply src/Utilities/ExcelImporterUtils.cs]: Don't know for sure, but removed it anyway

File: src/Utilities/ExcelImporterUtils.cs (hunk 1 above)
Comment (redbaty) [reply src/Utilities/ExcelImporterUtils.cs]: namespace is reserved, I renamed it to nameSpace

File: src/Utilities/ExcelImporterUtils.cs (hunk 3 above)
Comment (hughbe) [reply src/Utilities/ExcelImporterUtils.cs]: True, this makes sense. Still, let's only return `IEnumerable<ExcelClassMap>` as we don't want to leak the implementation

File: src/Utilities/ExcelImporterUtils.cs (hunk 1 above)
Comment (hughbe) [reply src/Utilities/ExcelImporterUtils.cs]: I updated this to be `namespaceString` - `nameSpace` is a bit weird IMO

**Comment by redbaty** [comment]
@hughbe It also misses Assembly.GetTypes, I'll see if I find a nuget package that fixes it

**Comment by redbaty** [comment]
@hughbe Done, it now supports 1.4

**Comment by redbaty** [comment]
Tests works on older XUnit version

File: src/Utilities/ExcelImporterUtils.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            if (namespaceString.Length == 0)
+            {
+                throw new ArgumentException("The namespace cannot be empty.", nameof(namespaceString));
+            }
+
+            var classes = assembly
+                .GetTypes()
+                .Where(t => t.IsAssignableFrom(typeof(ExcelClassMap)) && t.Namespace == namespaceString);
+
+            var objects = classes.Select(Activator.CreateInstance).OfType<ExcelClassMap>().ToList();
```
Comment (hughbe) [inline src/Utilities/ExcelImporterUtils.cs]: Do you think we should throw an error if there are no objects in the list?

Comment (redbaty) [reply src/Utilities/ExcelImporterUtils.cs]: Sure thing, it helps alot with debugging

**Comment by hughbe** [comment]
Thanks for working this through! Good to see this merged :)

**Comment by redbaty** [comment]
@hughbe Hey we should update the Readme.md to include this, let me know if you want me to do it! 😄

**Comment by hughbe** [comment]
Good idea - send in a Pr :)

**Pull request merged**